lib/fetch/index.js

```diff
@@ -740,34 +742,86 @@ function finalizeResponse (fetchParams, response) {
-  //    process response consume given response and nullOrBytes.on.
-  //    TODO
-  //    2. Let processBodyError be this step: run fetchParams’s process
-  //    response consume given response and failure.on.
-  //    TODO
-  //    3. If response’s body is null, then queue a fetch task to run
-  //    processBody given null, with fetchParams’s task destination.on.
-  //    TODO
-  //    4. Otherwise, fully read response’s body given processBody,
-  //    processBodyError, and fetchParams’s task destination.on.
-  //    TODO
-
-  // TODO (spec): The spec doesn't specify this but we need to
-  // terminate fetch if we have an error response.
-  if (response.type === 'error') {
-    context.terminate({ reason: response.error })
+  // 4. If response’s body is null, then run processResponseEndOfBody.
+  if (response.body == null) {
+    processResponseEndOfBody()
+  }
+
+  // 5. Otherwise:
+  else {
+    TransformStream ??= require('stream/web').TransformStream
+    // 1. Let transformStream be a new a TransformStream.
+
+    // 2. Let identityTransformAlgorithm be an algorithm which, given chunk, enqueues chunk in transformStream.
+    const identityTransformAlgorithm = (chunk, controller) => {
+      controller.enqueue(chunk)
+    }
+
+    const transformStream = new TransformStream({
+      // 3. Set up transformStream with transformAlgorithm set to identityTransformAlgorithm and
+      //    flushAlgorithm set to processResponseEndOfBody.
+      transform: identityTransformAlgorithm,
+      flush: processResponseEndOfBody
+    })
+
+    // 4. Set response’s body to the result of piping response’s body through transformStream.
+    // Khafra note: although the spec is unclear, only readable streams can be piped through
+    if (response.body != null && isReadable(response.body?.stream)) {
+      response.body = response.body.pipeThrough(transformStream)
+    }
+  }
+
+  // 6. If fetchParams’s process response consume body is non-null, then:
+  if (fetchParams.processResponseConsumeBody != null) {
+    // 1. Let processBody given nullOrBytes be this step: run fetchParams’s process response consume body given response and nullOrBytes.
+    const processBody = (nullOrBytes) => {
+      fetchParams.processResponseConsumeBody(response, nullOrBytes)
+    }
+
+    // 2. Let processBodyError be this step: run fetchParams’s process response consume body given response and failure.
+    const processBodyError = (failure) => {
+      fetchParams.processResponseConsumeBody(response, failure)
+    }
+
+    // 3. If response’s body is null, then queue a fetch task to run processBody given null, with fetchParams’s task destination.
+    if (response.body == null) {
+      processBody(null)
+    }
+
+    // 4. Otherwise, fully read response’s body given processBody, processBodyError, and fetchParams’s task destination.
+    else {
+      fullyReadBody(response.body, processBody, processBodyError, fetchParams.taskDestination)
```

Comment:
```suggestion
  await response.body.arrayBuffer().then(processBody, processBodyError)
```

I think this might be sufficient?

Reply:
If we do this, I can remove fullyReadBody, fullyReadBodyAsPromise, and ReadableStreamDefaultReaderReadAllBytes. The down side is that we aren't following the spec verbatim.

Is this a fair trade off? @mcollina @ronag

Reply:
Done in bd4642f2bb78131c25f5e21ec72b314faae09bf6!